### PR TITLE
[RF] Translate RooStats tutorial rs101_limitexample.C to Python

### PR DESCRIPTION
**Pull request by busorgin**
# This Pull request:

## Changes or fixes:
Add rs101_limitexample.py - Converted from rs101_limitexample.C

## Checklist:

- [ ] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes # Part of [8758](https://github.com/root-project/root/issues/8758)

### CODE REVIEW

File: tutorials/roostats/rs101_limitexample.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+## \notebook
+## Limits: number counting experiment with uncertainty on both the background rate and signal efficiency.
+##
+## The usage of a Confidence Interval Calculator to set a limit on the signal is illustrated
+##
+## \macro_image
+## \macro_output
+## \macro_code
+##
+## \author Kyle Cranmer
```
Comment (guitargeek) [inline tutorials/roostats/rs101_limitexample.py:12]:
```suggestion
## \date June 2022
## \authors Artem Busorgin, Kyle Cranmer (C++ version)
```
Please add also the date, and you can also add your name here to get some credits :)

File: tutorials/roostats/rs101_limitexample.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+# 3-d plot of the parameter points
+dataCanvas.cd(2)
+
+# also plot the points in the markov chain
+chainData = mcInt.GetChainAsDataSet()
+
+assert chainData
+print("plotting the chain data - nentries = ", chainData.numEntries())
+chain = ROOT.RooStats.GetAsTTree("chainTreeData", "chainTreeData", chainData)
+assert chain
```
Comment (guitargeek) [inline tutorials/roostats/rs101_limitexample.py:159]:
```suggestion
```
No asserts in tutorials

File: tutorials/roostats/rs101_limitexample.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+print("plotting the scanned points used in the frequentist construction - npoints = ", parScanData.numEntries())
+
+gr = ROOT.TGraph2D(parScanData.numEntries())
+for ievt in range(parScanData.numEntries()):
+    evt = parScanData.get(ievt)
+    x = evt.getRealValue("ratioBkgEff")
+    y = evt.getRealValue("ratioSigEff")
+    z = evt.getRealValue("s")
+    gr.SetPoint(ievt, x, y, z)
+    # print("{}  {}  {}  {}".format(ievt, x, y, z))
```
Comment (guitargeek) [inline tutorials/roostats/rs101_limitexample.py:177]:
```suggestion
```
It's not good to have commented out code in tutorials.

File: tutorials/roostats/rs101_limitexample.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+gSigBkg = wspace.var("gSigBkg") # global obs for background efficiency
+gSigEff.setConstant()
+gSigBkg.setConstant()
+
+# Create an example dataset with 160 observed events
+obs.setVal(160.)
+data = ROOT.RooDataSet("exampleData", "exampleData", ROOT.RooArgSet(obs))
+data.add(obs)
+
+all = ROOT.RooArgSet(s, ratioBkgEff, ratioSigEff)
```
Comment (guitargeek) [inline tutorials/roostats/rs101_limitexample.py:56]:
```suggestion
```
`all` is never used. You should delete this line

File: tutorials/roostats/rs101_limitexample.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+               "sum(s[50,0,120]*ratioSigEff[1.,0,3.],b[100]*ratioBkgEff[1.,0.,3.]))") # counting model
+# wspace.factory("Gaussian::sigConstraint(ratioSigEff,1,0.05)") # 5% signal efficiency uncertainty
+# wspace.factory("Gaussian::bkgConstraint(ratioBkgEff,1,0.1)") #  10% background efficiency uncertainty
+wspace.factory("Gaussian::sigConstraint(gSigEff[1,0,3],ratioSigEff,0.05)") # 5% signal efficiency uncertainty
+wspace.factory("Gaussian::bkgConstraint(gSigBkg[1,0,3],ratioBkgEff,0.2)") # 10% background efficiency uncertainty
+wspace.factory("PROD::modelWithConstraints(countingModel,sigConstraint,bkgConstraint)") # product of terms
+wspace.Print()
+
+modelWithConstraints = wspace.pdf("modelWithConstraints") # get the model
+obs = wspace.var("obs") # get the observable
```
Comment (guitargeek) [inline tutorials/roostats/rs101_limitexample.py:37]:
```suggestion
modelWithConstraints = wspace["modelWithConstraints"] # get the model
obs = wspace["obs"] # get the observable
```

In PyRoot, you can universally get anything from a RooWorkspace with the `[]`-operator. You should use this everytime you see `var()`, `pdf()`, or `function()` in C++ (there are also more occurences of `var()` to replace in the lines below).

File: tutorials/roostats/rs101_limitexample.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+constrainedParams = ROOT.RooArgSet(ratioSigEff, ratioBkgEff) # need to constrain these in the fit (should change default behavior)
+
+gSigEff = wspace.var("gSigEff") # global observables for signal efficiency
+gSigBkg = wspace.var("gSigBkg") # global obs for background efficiency
+gSigEff.setConstant()
+gSigBkg.setConstant()
+
+# Create an example dataset with 160 observed events
+obs.setVal(160.)
+data = ROOT.RooDataSet("exampleData", "exampleData", ROOT.RooArgSet(obs))
```
Comment (guitargeek) [inline tutorials/roostats/rs101_limitexample.py:53]:
```suggestion
data = ROOT.RooDataSet("exampleData", "exampleData", {obs})
```

**Review by guitargeek — CHANGES_REQUESTED** [review]
Hi @busorgin thank you so much for this contribution, which is really appreciated :+1: I which context are you doing this? Are you a RooFit user who wants to contribute, or you were just browsing GitHub for issues you could address?

Please address the requests I have (mostly about Pythonic code, and one fix for the destruction order fiasco).

**Comment by guitargeek** [comment]
@phsft-bot build

**Comment by guitargeek** [comment]
Alright, thank you very much! That's very nice you chose ROOT to contribute to!

If you need any guidance on what you could help with, just ask me. Sometimes it's not easy to find the issues that are actually easy to work on, as the code is quite complex.

Are you looking only Python-related issues, or would you also be interested in doing some C++ contributions to beef up your C++ skills and experience?

As for this PR, I'll merge it if the CI bot tests pass.

**Comment by busorgin** [comment]
Thanks, I'll keep that in mind.

I was looking for both C++ and Python. I found #9859, seemed like a good starter item on the C++ side. Open to suggestions if there are other good issues to look into.

**Comment by guitargeek** [comment]
Hi @busorgin, okay sounds great! The issue you linked is indeed a very good one to get started with RooFit C++ development. If you have any questions about it, we can follow up in the issue thread there.

Of course, it would also be of great help if you continue doing these RooStats tutorial translations! People use more Python than C++ now, and a better coverage of the Python tutorials is something requested by many users, in particular for RooStats. So among all the "easier" RooFit issues, the RooStats tutorial translations have the highest priority.

**Pull request merged**